test/unit/search/kmer_index/shape_iterator_tests.cpp

```diff
@@ -0,0 +1,124 @@
+// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE
+// -----------------------------------------------------------------------------------------------------
+
+#include "gtest/gtest.h" // für googletest
+
+#include <algorithm>
+#include <cassert>
+#include <iostream>
+#include <string>
+#include <vector>
+
+#include <seqan3/alphabet/all.hpp>
+#include <seqan3/io/stream/debug_stream.hpp>
+#include <seqan3/search/kmer_index/shape.hpp>
+#include <seqan3/search/kmer_index/shape_iterator.hpp>
+#include <seqan3/range/view/take.hpp>
+
+using namespace seqan3;
+
+// ============================================================================
+//  shape
+// ============================================================================
+
+TEST(ShapeTest, HandlesSizes)
+{
+    EXPECT_EQ(std::ranges::size(shape()),    0u);
+    EXPECT_EQ(std::ranges::size(shape(0)),   0u);
+    EXPECT_EQ(std::ranges::size(shape(1)),   1u);
+    EXPECT_EQ(std::ranges::size(shape(31)), 31u);
+    EXPECT_EQ(std::ranges::size(shape(0,0)), 2u);
+    EXPECT_EQ(std::ranges::size(shape(1,1)), 2u);
+    EXPECT_EQ(std::ranges::size(shape(0,0,0,0,0,0)), 6u);
+    EXPECT_EQ(std::ranges::size(shape(1,0,1,0,1,1)), 6u);
+    EXPECT_EQ(std::ranges::size(shape(1,1,1,1,1,1)), 6u);
+}
+
+// size 3 ausreichend?
+TEST(ShapeTest, HandlesShapes)
+{
+    shape s0(0,0,0);
+    std::array<bool,3> a0{false,false,false};
+    for (size_t i(0); i < a0.size(); i++)
+    {
+        EXPECT_EQ(s0.get()[i],a0[i]);
+    }
+
+    shape s1(1,0,1);
+    std::array<bool,3> a1{true,false,true};
+    for (size_t i(0); i < a1.size(); i++)
+    {
+        EXPECT_EQ(s1.get()[i],a1[i]);
+    }
+
+    shape s2(3);
+    std::array<bool,3> a2{true,true,true};
+    for (size_t i(0); i < a2.size(); i++)
+    {
+        EXPECT_EQ(s2.get()[i],a2[i]);
+    }
+}
+
+// ============================================================================
+//  iterator
+// ============================================================================
+
+
+TEST(ShapeIteratorTest, AlphabetConstructor)
+{
+    shape_iterator it(("C"_dna4).begin(), shape(1));
+
+    EXPECT_NE(&it, nullptr);
+    EXPECT_EQ(*it,1u);
+}
+
+TEST(ShapeIteratorTest, CalculateHashByConstruction)
+{
+    shape_iterator it(("ACGT"_dna4).begin(), shape(4));
+
+    EXPECT_EQ(*it, 27u);
+}
+
+
+TEST(ShapeIteratorTest, RollingHash)
+{
+
+    std::array<size_t,3> expected_hashes{1,6,11};
+    dna4_vector dna4_sample{"ACGT"_dna4};
+
+    shape shape2(2);
+    shape_iterator it(dna4_sample.begin(), shape2);
+
+    EXPECT_EQ(*it,expected_hashes[0]);
+    ++it; // *(++it) funktioniert nicht. warum?
```

Comment:
`*it; ++it` != `*++it`
`*++it` == `++it; *it`
`*it; ++it` == `*it++`, but produces a copy

Reply:
I just couldn't compile *++it. Postfix operator is not defined in shape_iterator.
`
EXPECT_EQ(*it,expected_hashes[1]);
    ++it;
`
could not be written as
`EXPECT_EQ(*++it,expected_hashes[1]);`
Compiler says:

```
error: void value not ignored as it ought to be
     EXPECT_EQ(*++it,expected_hashes[1]);
```
I guess it would be more neat, that's all.

Reply:
I mean the main point here is, that comparing `*++it` is semantically wrong since you first want to compare the value and then increment `it`. `*++it` would first increment `it` *and then* dereference it. Hence you'd be comparing the wrong values.

You could also implement a postincrement, but this would also have to create a copy.

Anyway, having the `++` on the next line is more readable, so I'd leave it like that.

Reply:
I was aware of that. I wanted the increment in the last two calls, not in the first one. Well, in that case, I'll leave it as it is.

Reply:
Ah 💡 

Your `operator++` should look like this: `shape_iterator & operator++()` and have a `return *this;` at the end.
Then it works.

Reply:
yes, that makes sense, thank you!